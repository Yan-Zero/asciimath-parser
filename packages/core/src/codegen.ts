import type { ChildNode, MatrixNode, RootNode } from './parser'
import { AlignDirection, NodeTypes } from './parser'

function getArrayBoundary(node: MatrixNode) {
  const div = node.dividerIndices
  let beginArray = '\\begin{array}'
  let ch = 'c'
  if (node.alignment === AlignDirection.Left)
    ch = 'l'

  if (div.length) {
    const divMax = div[div.length - 1]
    for (let i = div.length - 1; i >= 1; i--)
      div[i] -= div[i - 1]

    beginArray += '{'
    for (let i = 0; i < div.length; i++)
      beginArray += `${''.padEnd(div[i], ch)}|`

    // MathJax would complain if the array env arg
    // is not consistent with the elements of the matrix.
    const maxCol = Math.max(...node.params.map(i => i.length))
    beginArray += `${''.padEnd(maxCol - divMax, ch)}}`
  }
  else {
    const maxCol = Math.max(...node.params.map(c => c.length))
    beginArray += '{' + ''.padEnd(maxCol, ch) + '}'
  }
  return [beginArray, '\\end{array}']
}

function codegen(node: ChildNode | RootNode): string {
  switch (node.type) {
    case NodeTypes.Const: {
      return node.tex
    }
    case NodeTypes.Root: {
      let res = node.body.map(codegen).join(' ')
      if (node.body.find(n => n.type === NodeTypes.Const && n.value === '&'))
        res = `\\begin{aligned}${res}\\end{aligned}`
      return res
    }
    case NodeTypes.Flat: {
      return node.body.map(codegen).join(' ')
    }
    case NodeTypes.Matrix: {
      // const [beginMatrix, endMatrix] = getMatrixBoundary(node)
      const [arrayBegin, arrayEnd] = getArrayBoundary(node)
      return [
        node.lparen,
        // beginMatrix,
        arrayBegin,
        node.params.map(i => i.map(codegen).join('&')).join('\\\\'),
        '\\\\',
        arrayEnd,
        // endMatrix,
        node.rparen,
      ].join(' ')
    }
    case NodeTypes.ParamOne: {
      return node.tex.replace('$1', codegen(node.params))
    }
    case NodeTypes.ParamTwo: {
      return node.tex.replace('$1', codegen(node.params[0])).replace('$2', codegen(node.params[1]))
    }
  }
}

export {
  codegen,
}
