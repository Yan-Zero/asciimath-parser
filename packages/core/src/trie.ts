import { SYMBOLMAP, TokenTypes } from './symbols'

interface TokenizedValue {
  value: string
  isKeyWord: boolean
  current: number
  tex: string
  type: TokenTypes
}

const NUMBERPATTERN = /[0-9]/
const STRINGPATTERN = /\S/

type GeneTokenFn = (config: {
  current: number
  value?: string
}) => TokenizedValue

const createTextToken: GeneTokenFn = (config: {
  current: number
  value?: string
}) => {
  const { value = '', current } = config
  return { value, isKeyWord: false, current, tex: value, type: TokenTypes.Text }
}

const createConstToken: GeneTokenFn = (config: {
  current: number
  value?: string
}) => {
  const { value = '', current } = config
  return { value, isKeyWord: false, current, tex: value, type: TokenTypes.Const }
}

class Trie {
  private _root: TrieNode
  private _char_to_index: Map<string, number> = new Map()
  private _n: number

  public constructor(nodes: string[]) {
    if (nodes.length === 0)
      throw new Error('Cannot create Trie since the length of nodes is 0')

    nodes.forEach((n) => {
      if (n.length !== 1)
        throw new Error(`Value \`${n}\` is invalid, the length of char must be 1`)
    })

    const dedupped = Array.from(new Set(nodes))

    this._n = dedupped.length
    this._root = new TrieNode(this._n)
    dedupped.forEach((ch, i) => {
      this._char_to_index.set(ch, i)
    })
  }

  private c2i(key: string) {
    return this._char_to_index.get(key)
  }

  public insert(key: string) {
    if (key.length === 0)
      return

    let root = this._root

    ;([...key]).forEach((ch, i) => {
      const idx = this.c2i(ch)
      if (typeof idx === 'undefined')
        throw new Error(`key \`${ch}\` not in key set`)
      if (root._nextNode[idx] === null)
        root._nextNode[idx] = new TrieNode(this._n)
      root = root._nextNode[idx]!
      if (i === key.length - 1)
        root._end = true
    })
  }

  public search(word: string): boolean {
    if ((!this._root._nextNode.find(i => i !== null)) || word.length === 0)
      return false
    let root = this._root
    let i = 0
    for (; i < word.length; i++) {
      const ch = word[i]
      const idx = this.c2i(ch)
      if (typeof idx === 'undefined')
        throw new Error(`key \`${ch}\` not in key set`)
      if (root._nextNode[idx] === null)
        return false
      root = root._nextNode[idx]!
    }
    if (i === word.length)
      return true
    return false
  }

  /**
   * @param letters input string
   * @returns value: matched keyword
   * current: the string cursor
   */
  public tryParsing(letters: string[], start = 0): TokenizedValue {
    let value = ''
    let root = this._root
    let isKeyWord = false
    let current = start
    // let depth = 0
    for (; current < letters.length; current++) {
      const ch = letters[current]
      // console.log('depth', depth, ch)

      // the spaces are processed in `tryParsingAll`
      // if (/\s/.test(ch)) {
      //   current++
      //   break
      // }

      const idx = this.c2i(ch)
      // cannot find key in symbol map
      if (typeof idx === 'undefined')
        break

      if (root._nextNode[idx] === null)
        break

      value += ch
      root = root._nextNode[idx]!
      isKeyWord = root._end
      // depth++
    }
    const ret = (() => {
      if (isKeyWord)
        return SYMBOLMAP.get(value)!
      return { tex: value, type: TokenTypes.StringLiteral }
    })()
    return { value, isKeyWord, current, ...ret }
  }

  public tryParsingNumber(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (NUMBERPATTERN.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    if (ch === '.') {
      value += ch
      ch = letters[++current]
    }
    while (NUMBERPATTERN.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    return { value, isKeyWord: false, current, tex: value, type: TokenTypes.NumberLiteral }
  }

  public tryParsingString(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (STRINGPATTERN.test(ch) && current < letters.length) {
      const idx = this.c2i(ch)
      if (typeof idx !== 'undefined' && this._root._nextNode[idx] !== null)
        break
      value += ch
      ch = letters[++current]
    }
    return { value, isKeyWord: false, current, tex: value, type: TokenTypes.StringLiteral }
  }

  public tryParsingNewLines(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (/\n/.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    if (value.length >= 2)
      return { value, isKeyWord: true, current, tex: '\\\\', type: TokenTypes.Align }
    else
      return { value: '', isKeyWord: false, current, tex: '', type: TokenTypes.None }
  }

  private getPlainTextInDoubleQuote(letters: string[], current: number): { value: string; current: number } {
    let value = ''
    let ch = letters[current]
    if (ch === '"') {
      ch = letters[++current]
      while (ch !== '"' && current < letters.length) {
        value += ch
        ch = letters[++current]
      }
      if (ch === '"') {
        current++
        return { current, value }
      }
    }
    return { value, current }
  }

  public tryParsingText(letters: string[], current: number): TokenizedValue {
    const { value, current: newCurrent } = this.getPlainTextInDoubleQuote(letters, current)
    return { value, isKeyWord: false, current: newCurrent, tex: value, type: TokenTypes.Text }
  }

  private getColorString(letters: string[], current: number): TokenizedValue {
    let color = ''
    while (current < letters.length) {
      const ch = letters[current]
      if (!/[#\da-z]/i.test(ch))
        break
      color += ch
      current++
    }

    return { value: color, isKeyWord: false, current, tex: color, type: TokenTypes.Const }
  }

  private skipSpaces(letters: string[], current: number): number {
    while (current < letters.length) {
      const ch = letters[current]
      if (!/\s/.test(ch))
        break
      current++
    }
    return current
  }

  private processColor(letters: string[], current: number): TokenizedValue {
    let existParen = false
    let res = { value: '', isKeyWord: false, current, tex: '', type: TokenTypes.Const }
    current = this.skipSpaces(letters, current)
    if (current >= letters.length)
      return res
    { // detect paren
      const ch = letters[current]
      if (/[\(\{\[]/.test(ch)) {
        existParen = true
        current++
      }
    }
    current = this.skipSpaces(letters, current)
    res = this.getColorString(letters, current)
    current = res.current
    current = this.skipSpaces(letters, current)
    res.current = current
    if (current >= letters.length)
      return res
    {
      const ch = letters[current]
      if (/[\)\}\]]/.test(ch) && existParen)
        current++
    }
    res.current = current
    return res
  }

  private getPlainText(letters: string[], current: number, fn: GeneTokenFn): TokenizedValue {
    let useParen = false
    let ch = letters[current]
    while (/\s/.test(ch))
      ch = letters[++current]

    useParen = ch === '('
    let value = ''
    if (useParen) {
      ch = letters[++current]
      while (current < letters.length && ch !== ')') {
        value += ch
        ch = letters[++current]
      }
      current++
      return fn({ current, value })
    }
    // do not use paren, then read a word
    while (current < letters.length && /\S/.test(ch)) {
      value += ch
      ch = letters[++current]
    }
    return fn({ current, value })
  }

  public tryParsingAll(word: string) {
    let current = 0
    const tokens: TokenizedValue[] = []
    let counter = 0
    const letters = [...word]
    while (current < letters.length) {
      {
        const t = this.tryParsingNewLines(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process spaces
      if (/\s/.test(letters[current])) {
        current++
        continue
      }
      // process potential keywords
      let t = this.tryParsing(letters, current)
      current = t.current
      if (t.value !== '') {
        switch (t.value) {
          case 'text': {
            t = this.getPlainText(letters, current, createTextToken)
            current = t.current
            tokens.push(t)
            break
          }
          case 'tex': {
            // t = this.getPlainText(letters, current, createConstToken)
            current = this.skipSpaces(letters, current)
            const { value: v, current: c } = this.getPlainTextInDoubleQuote(letters, current)
            current = c
            t = createConstToken({ current, value: v })
            tokens.push(t)
            break
          }
          case 'hspace': {
            tokens.push(t)
            const nt = this.getPlainText(letters, current, createConstToken)
            current = nt.current
            tokens.push(nt)
            break
          }
          case 'color': {
            tokens.push(t)
            const nt = this.processColor(letters, current)
            current = nt.current
            tokens.push(nt)
            break
          }
          default: {
            tokens.push(t)
          }
        }
        continue
      }
      // process numbers
      {
        const t = this.tryParsingNumber(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process text wrapped with double quote
      {
        const t = this.tryParsingText(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process other string
      {
        const t = this.tryParsingString(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      counter++
      if (counter > letters.length * 2)
        throw new Error('Oops! There may be an infinity loop')
    }
    return tokens
  }
}

class TrieNode {
  _nextNode: (TrieNode | null)[] = []
  _end = false
  constructor(n: number) {
    this._nextNode = Array.from({ length: n }, () => null)
  }
}

function createTrie(config: {
  extConst?: Array<[string, string]>
} = {}) {
  const charset: Set<string> = new Set([])
  config.extConst?.forEach(([k, v]) => {
    SYMBOLMAP.set(k, { type: TokenTypes.Const, tex: v })
  })
  for (const k of SYMBOLMAP.keys())
    [...k].forEach(i => charset.add(i))
  const chars = Array.from(charset)
  chars.push(' ')

  const trie = new Trie(chars)
  for (const k of SYMBOLMAP.keys())
    trie.insert(k)

  return trie
}

export {
  Trie,
  TrieNode,
  createTrie,
  TokenizedValue,
}
