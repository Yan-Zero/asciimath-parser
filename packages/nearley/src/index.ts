import Nearley from 'nearley'
import initGrammar from './grammar.js'
import initLexer from './lexer.js'
import type { Symbols } from './symbols.js'
import initSymbols from './symbols.js'
import initGenerator from './to-tex.js'

export type Ast = any
type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * @default false
   * throws error on parse
   */
  throws?: boolean
  /**
   * Add custom symbols.
   *
   * For example:
   * {
   *   keyword: {
   *     dx: { tex: '{\\text{d}x}' },
   *     dy: { tex: '{\\text{d}y}' },
   *     dz: { tex: '{\\text{d}z}' },
   *     dt: { tex: '{\\text{d}t}' },
   *     ee: { tex: '\\text{e}' },
   *     ii: { tex: '\\text{i}' },
   *   }
   * }
   */
  symbols?: Symbols
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  return {
    display: config?.display ?? true,
    throws: config?.throws ?? false,
    symbols: {
      keyword: {
        dx: { tex: '{\\text{d}x}' },
        dy: { tex: '{\\text{d}y}' },
        dz: { tex: '{\\text{d}z}' },
        dt: { tex: '{\\text{d}t}' },
        ee: { tex: '\\text{e}' },
        ii: { tex: '\\text{i}' },
      },
      ...config?.symbols,
    },
    replaceBeforeTokenizing: [
      // html entity
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
      ...(config?.replaceBeforeTokenizing || []),
    ],
  }
}

class AsciiMath {
  public display: boolean
  public throws: boolean
  public replaceBeforeTokenizing: ReplaceLaw[]
  public lexer: Nearley.Lexer
  public parser: Nearley.Parser
  private genTex: ((ast: Ast) => string)
  private initState: { [key: string]: any; lexerState: Nearley.LexerState }
  constructor(config?: AsciiMathConfig) {
    const { display, throws, symbols, replaceBeforeTokenizing } = resolveConfig(config)
    this.display = display
    this.throws = throws
    this.replaceBeforeTokenizing = replaceBeforeTokenizing
    const allSymbols = initSymbols(symbols)
    const lexer = this.lexer = initLexer(allSymbols)
    const grammar = initGrammar(lexer)
    const compiledGrammar = Nearley.Grammar.fromCompiled(grammar)
    this.parser = new Nearley.Parser(compiledGrammar)
    this.initState = this.parser.save()
    this.genTex = initGenerator(allSymbols)
  }

  toTex(code: string): string {
    this.parser.restore(this.initState)
    try {
      code = this.replaceBeforeTokenizing.reduce((prev, curLaw) => {
        // nodejs 14 没有 replaceAll 方法, 不要用 replaceAll
        return prev.replace(new RegExp(curLaw[0], 'g'), curLaw[1] as any)
      }, code)
      this.parser.feed(code)
      if (this.parser.results.length === 0) {
        throw new Error('unexpected end of input')
      }
      else if (this.parser.results.length > 1) {
        console.error(this.parser.results)
        throw new Error('ambiguous parse')
      }
      let res = this.genTex(this.parser.results)
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      if (this.throws)
        throw e
      console.error(e)
      const message = String(e)
      let index = message.indexOf(' Instead, I was expecting to see one of the following:')
      if (index > -1) {
        return this.genTex({
          type: 'opOA',
          value: 'verb',
          $1: { value: message.slice(0, index) },
        })
      }

      index = message.indexOf(':\n')
      const end = index > -1 ? index : undefined
      return `\\text{${message.slice(0, end)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
