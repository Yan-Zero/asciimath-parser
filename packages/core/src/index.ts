import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
