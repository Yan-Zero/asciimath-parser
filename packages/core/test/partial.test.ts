import { describe, expect, it } from 'vitest'
import { createTrie } from '../src/trie'
import { AsciiMath } from '../src'
import { parser } from '../src/parser'
import { codegen } from '../src/codegen'

describe('partial and derivative', () => {
  it('should tokenize expressions', () => {
    const code = 'part f x'
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    expect(tokens).toMatchSnapshot()
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
    expect(codegen(ast)).toMatchSnapshot()
  })
  it('should generate partial expressions', () => {
    const am = new AsciiMath({ display: false })
    expect(am.toTex('part f x = 2x'))
      .toMatchInlineSnapshot('"\\\\frac{ \\\\partial f }{ \\\\partial x } = 2 x"')
    expect(am.toTex('part^3 f (x y^2) = 2x^2 y'))
      .toMatchInlineSnapshot('"\\\\frac{ \\\\partial ^{ 3 } f }{ \\\\partial x \\\\partial y ^{ 2 } } = 2 x ^{ 2 } y"')
    expect(am.toTex('part^3 f x = 3x^2'))
      .toMatchInlineSnapshot('"\\\\frac{ \\\\partial ^{ 3 } f }{ \\\\partial x ^{ 3 } } = 3 x ^{ 2 }"')
    expect(am.toTex('dd^3 f x = 3x^2'))
      .toMatchInlineSnapshot('"\\\\frac{ \\\\mathrm{d} ^{ 3 } f }{ \\\\mathrm{d} x ^{ 3 } } = 3 x ^{ 2 }"')
  })
  it('should parse x^2', () => {
    const code = 'part^3 f (x y^2)'
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
  })

  it('should parse matrix', () => {
    const code = '[part^2 f x, part^2 f (x y); part^2 f (y x), part^2 f y]'
    const trie = createTrie()
    const tokens = trie.tryParsingAll(code)
    expect(tokens).toMatchSnapshot()
    const ast = parser(tokens)
    expect(ast).toMatchSnapshot()
    expect(codegen(ast)).toMatchSnapshot()
  })
})
