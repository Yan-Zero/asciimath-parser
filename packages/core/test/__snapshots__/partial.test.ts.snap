// Vitest Snapshot v1

exports[`partial and derivative > should parse matrix 1`] = `
[
  {
    "current": 1,
    "isKeyWord": true,
    "tex": "[",
    "type": "LParen",
    "value": "[",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "\\\\partial",
    "type": "OperatorPartial",
    "value": "part",
  },
  {
    "current": 6,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 11,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 12,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 17,
    "isKeyWord": true,
    "tex": "\\\\partial",
    "type": "OperatorPartial",
    "value": "part",
  },
  {
    "current": 18,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 19,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 21,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 23,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 24,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 26,
    "isKeyWord": false,
    "tex": "y",
    "type": "StringLiteral",
    "value": "y",
  },
  {
    "current": 27,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 28,
    "isKeyWord": true,
    "tex": ";",
    "type": "Split",
    "value": ";",
  },
  {
    "current": 33,
    "isKeyWord": true,
    "tex": "\\\\partial",
    "type": "OperatorPartial",
    "value": "part",
  },
  {
    "current": 34,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 35,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 37,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 39,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 40,
    "isKeyWord": false,
    "tex": "y",
    "type": "StringLiteral",
    "value": "y",
  },
  {
    "current": 42,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 43,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 44,
    "isKeyWord": true,
    "tex": ",",
    "type": "Split",
    "value": ",",
  },
  {
    "current": 49,
    "isKeyWord": true,
    "tex": "\\\\partial",
    "type": "OperatorPartial",
    "value": "part",
  },
  {
    "current": 50,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorSup",
    "value": "^",
  },
  {
    "current": 51,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 53,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 55,
    "isKeyWord": false,
    "tex": "y",
    "type": "StringLiteral",
    "value": "y",
  },
  {
    "current": 56,
    "isKeyWord": true,
    "tex": "]",
    "type": "RParen",
    "value": "]",
  },
]
`;

exports[`partial and derivative > should parse matrix 2`] = `
{
  "body": [
    {
      "alignment": "c",
      "dividerIndices": [],
      "lparen": "\\\\left[",
      "params": [
        [
          {
            "body": [
              {
                "params": [
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                      {
                        "tex": "f",
                        "type": "Const",
                        "value": "f",
                      },
                    ],
                    "type": "Flat",
                  },
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "x",
                        "type": "Const",
                        "value": "x",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                    ],
                    "type": "Flat",
                  },
                ],
                "tex": "\\\\frac{ $1 }{ $2 }",
                "type": "ParamTwo",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "params": [
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                      {
                        "tex": "f",
                        "type": "Const",
                        "value": "f",
                      },
                    ],
                    "type": "Flat",
                  },
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "x",
                        "type": "Const",
                        "value": "x",
                      },
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "y",
                        "type": "Const",
                        "value": "y",
                      },
                    ],
                    "type": "Flat",
                  },
                ],
                "tex": "\\\\frac{ $1 }{ $2 }",
                "type": "ParamTwo",
              },
            ],
            "type": "Flat",
          },
        ],
        [
          {
            "body": [
              {
                "params": [
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                      {
                        "tex": "f",
                        "type": "Const",
                        "value": "f",
                      },
                    ],
                    "type": "Flat",
                  },
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "y",
                        "type": "Const",
                        "value": "y",
                      },
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "x",
                        "type": "Const",
                        "value": "x",
                      },
                    ],
                    "type": "Flat",
                  },
                ],
                "tex": "\\\\frac{ $1 }{ $2 }",
                "type": "ParamTwo",
              },
            ],
            "type": "Flat",
          },
          {
            "body": [
              {
                "params": [
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                      {
                        "tex": "f",
                        "type": "Const",
                        "value": "f",
                      },
                    ],
                    "type": "Flat",
                  },
                  {
                    "body": [
                      {
                        "tex": "\\\\partial",
                        "type": "Const",
                        "value": "\\\\partial",
                      },
                      {
                        "tex": "y",
                        "type": "Const",
                        "value": "y",
                      },
                      {
                        "params": {
                          "tex": "2",
                          "type": "Const",
                          "value": "2",
                        },
                        "tex": "^{ $1 }",
                        "type": "ParamOne",
                      },
                    ],
                    "type": "Flat",
                  },
                ],
                "tex": "\\\\frac{ $1 }{ $2 }",
                "type": "ParamTwo",
              },
            ],
            "type": "Flat",
          },
        ],
      ],
      "rparen": "\\\\right]",
      "type": "Matrix",
    },
  ],
  "type": "Root",
}
`;

exports[`partial and derivative > should parse matrix 3`] = `"\\\\left[ \\\\begin{array}{cc} \\\\frac{ \\\\partial ^{ 2 } f }{ \\\\partial x ^{ 2 } } & \\\\frac{ \\\\partial ^{ 2 } f }{ \\\\partial x \\\\partial y } \\\\\\\\ \\\\frac{ \\\\partial ^{ 2 } f }{ \\\\partial y \\\\partial x } & \\\\frac{ \\\\partial ^{ 2 } f }{ \\\\partial y ^{ 2 } } \\\\end{array} \\\\right]"`;

exports[`partial and derivative > should parse x^2 1`] = `
{
  "body": [
    {
      "params": [
        {
          "body": [
            {
              "tex": "\\\\partial",
              "type": "Const",
              "value": "\\\\partial",
            },
            {
              "params": {
                "tex": "3",
                "type": "Const",
                "value": "3",
              },
              "tex": "^{ $1 }",
              "type": "ParamOne",
            },
            {
              "tex": "f",
              "type": "Const",
              "value": "f",
            },
          ],
          "type": "Flat",
        },
        {
          "body": [
            {
              "tex": "\\\\partial",
              "type": "Const",
              "value": "\\\\partial",
            },
            {
              "tex": "x",
              "type": "Const",
              "value": "x",
            },
            {
              "tex": "\\\\partial",
              "type": "Const",
              "value": "\\\\partial",
            },
            {
              "body": [
                {
                  "tex": "y",
                  "type": "Const",
                  "value": "y",
                },
                {
                  "params": {
                    "tex": "2",
                    "type": "Const",
                    "value": "2",
                  },
                  "tex": "^{ $1 }",
                  "type": "ParamOne",
                },
              ],
              "type": "Flat",
            },
          ],
          "type": "Flat",
        },
      ],
      "tex": "\\\\frac{ $1 }{ $2 }",
      "type": "ParamTwo",
    },
  ],
  "type": "Root",
}
`;

exports[`partial and derivative > should tokenize expressions 1`] = `
[
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "\\\\partial",
    "type": "OperatorPartial",
    "value": "part",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
]
`;

exports[`partial and derivative > should tokenize expressions 2`] = `
{
  "body": [
    {
      "params": [
        {
          "body": [
            {
              "tex": "\\\\partial",
              "type": "Const",
              "value": "\\\\partial",
            },
            {
              "tex": "f",
              "type": "Const",
              "value": "f",
            },
          ],
          "type": "Flat",
        },
        {
          "body": [
            {
              "tex": "\\\\partial",
              "type": "Const",
              "value": "\\\\partial",
            },
            {
              "tex": "x",
              "type": "Const",
              "value": "x",
            },
          ],
          "type": "Flat",
        },
      ],
      "tex": "\\\\frac{ $1 }{ $2 }",
      "type": "ParamTwo",
    },
  ],
  "type": "Root",
}
`;

exports[`partial and derivative > should tokenize expressions 3`] = `"\\\\frac{ \\\\partial f }{ \\\\partial x }"`;
